### PR TITLE
[transport] Add node ID to address mapping

### PR DESCRIPTION
**Pull request by gjc13**
This change introduces mechanism to map node ID to its address using mDNS.

Changes Include:

 #### Problem

Support mDNS discovery for operational address resolution
https://github.com/project-chip/connectedhomeip/issues/3121

 #### Summary of Changes

  - Rename mDNS publisher to DiscoveryManager.
  - Implement mDNS resolve in DiscoveryManager.
  - Modify SecureSessionManager and DeviceController to support using
    node ID only.
  - Publish mDNS service entries in linux sample app.
  - Fix Linux platform mDNS implementation bugs.
  - Test nodeID mapping in cirque.

### CODE REVIEW

**Review by andy31415 — APPROVED** [review]

File: src/lib/mdns/DiscoveryManager.cpp [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+#if CHIP_ENABLE_MDNS
+    DiscoveryManager * mgr = static_cast<DiscoveryManager *>(context);
+
+    if (mgr->mResolveDelegate == nullptr)
+    {
+        return;
+    }
+    if (error != CHIP_NO_ERROR)
+    {
+        ChipLogError(Discovery, "Node ID resovled failed with %s", chip::ErrorStr(error));
```
Comment (mspang) [inline src/lib/mdns/DiscoveryManager.cpp:283]: typo: resovled -> resolved

File: src/lib/mdns/DiscoveryManager.h [hunk 2]
```diff
@@ -45,30 +50,51 @@ class Publisher
...
      * This function stops publishing the device on mDNS.
      *
      */
     CHIP_ERROR StopPublishDevice();
 
+    /**
+     * This function registers the delegate to handle node id resolve results.
+     *
+     */
+    CHIP_ERROR RegisterResovleDelegate(ResolveDelegate * delegate);
```
Comment (mspang) [inline src/lib/mdns/DiscoveryManager.h:66]: typo: RegisterResovleDelegate -> RegisterResolveDelegate

File: src/transport/PeerConnections.h [hunk 3]
```diff
@@ -184,31 +198,37 @@ class PeerConnections
...
+     * @return the state found, nullptr if not found
      */
     CHECK_RETURN_VALUE
-    bool FindPeerConnectionState(Optional<NodeId> nodeId, uint16_t peerKeyId, PeerConnectionState ** state)
+    PeerConnectionState * FindPeerConnectionState(Optional<NodeId> nodeId, uint16_t peerKeyId, PeerConnectionState * begin)
     {
-        *state = nullptr;
-        for (size_t i = 0; i < kMaxConnectionCount; i++)
+        PeerConnectionState * state = nullptr;
+        PeerConnectionState * now   = &mStates[0];
```
Comment (mspang) [inline src/transport/PeerConnections.h:209]: I wouldn't use `now` as this suggests time. Maybe use `peer` and rename `state` to `matchingPeer`.

File: src/transport/SecureSessionMgr.cpp [hunk 4]
```diff
@@ -199,6 +213,40 @@ CHIP_ERROR SecureSessionMgrBase::NewPairing(const Optional<Transport::PeerAddres
...
+    }
+    else if (nodeId == kUndefinedNodeId && mCB != nullptr)
+    {
+        mCB->OnAddressResolved(CHIP_ERROR_UNKNOWN_RESOURCE_ID, kUndefinedNodeId, this);
+    }
+    else if (error == CHIP_NO_ERROR && nodeId != kUndefinedNodeId)
+    {
+        PeerConnectionState * state = nullptr;
+        PeerAddress addr;
+        bool hasAddressupdate = false;
```
Comment (mspang) [inline src/transport/SecureSessionMgr.cpp:230]: hasAddressupdate -> hasAddressUpdate

File: src/transport/SecureSessionMgr.cpp [hunk 5]
```diff
@@ -199,6 +213,40 @@ CHIP_ERROR SecureSessionMgrBase::NewPairing(const Optional<Transport::PeerAddres
...
+
+        // Though the spec says the service name is _chip._tcp, we are now supporting tcp for now.
```
Comment (mspang) [inline src/transport/SecureSessionMgr.cpp:232]: typo here: now -> not ?

**Review by rwalker-apple — APPROVED** [review]

**Review by BroderickCarlin — APPROVED** [review]

File: src/lib/mdns/DiscoveryManager.cpp [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+#include "platform/CHIPDeviceLayer.h"
+#include "support/CodeUtils.h"
+#include "support/ErrorStr.h"
+#include "support/RandUtils.h"
+
+#if CHIP_ENABLE_MDNS
+
+namespace {
+
+uint8_t HexToInt(char c)
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/lib/mdns/DiscoveryManager.cpp:34]: File https://github.com/project-chip/connectedhomeip/issues/3769

**Pull request merged**